Rewrote emscripten chapter to use root finding algo (#21)
NLESC-JCER/cpp2wasm: README.md

The binary format is stored as a WebAssembly module or `*.wasm` file, which can be loaded by all modern web browsers.

Instead of writing code in the WebAssembly language, there are compilers that can take C++/C code and compile it to wasm. [Emscripten](https://emscripten.org) is the most popular C++ to wasm compiler. Emscripten has been successfully used to port game engines like the Unreal engine to the browser making it possible to have complex 3D games in the browser without needing to install anything else than the web browser. To call C++ code (which has been compiled to wasm) from JavaScript, a binding is required. The binding will map C++ constructs to their JavaScript equivalent and back. The binding called [embind](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#embind) is declared in a C++ file which is included in the compilation.

> The binding of the algorithm will be

Reviewer: ```suggestion
The binding of the C++ code will be
```

```{.cpp file=src/wasm-newtonraphson.cpp}
// this C++ snippet is stored as src/wasm-newtonraphson.cpp
#include <emscripten/bind.h>

<<algorithm>>

using namespace emscripten;

EMSCRIPTEN_BINDINGS(newtonraphsonwasm) {
  class_<rootfinding::NewtonRaphson>("NewtonRaphson")
    .constructor<double>()
    .function("find", &rootfinding::NewtonRaphson::find)
    ;
}
```

The algorithm and binding can be compiled into a Web Assembly module with the Emscripten compiler called `emcc`.
To make live easier we configure the compile command to generate a `src/js/newtonraphsonwasm.js` file which exports the `createModule` function.

```{.awk #build-wasm}
emcc --bind -o src/js/newtonraphsonwasm.js -s MODULARIZE=1 -s EXPORT_NAME=createModule src/wasm-newtonraphson.cpp
```

The compilation also generates a `src/js/newtonraphsonwasm.wasm` file which will be loaded with the `createModule` function.

> The WebAssembly module must be loaded and initialized by calling the `createModule` function and waiting for the promise to resolve.

Reviewer: ```suggestion
The WebAssembly module must be loaded and initialized by calling the `createModule` function and waiting for the Javascript promise to resolve.
```

Author: done, but spelled JavaScript

```{.js #wasmpromise}
// this Javascript snippet is later referred to as wasmpromise
createModule().then((module) => {
  <<wasmcalc>>
});
```

The `module` variable contains the `NewtonRaphson` class we defined in the binding above.

The root finder can be called with.

```{.js #wasmcalc}
// this Javascript snippet is later referred to as wasmcalc
const epsilon = 0.001;
const finder = new module.NewtonRaphson(epsilon);
const guess = -20;
const root = finder.find(guess);
```

Append the root answer to the html page using [document manipulation functions](https://developer.mozilla.org/en-US/docs/Web/API/ParentNode/append).

```{.js #wasmcalc}
const answer = document.createElement('span');
answer.id = 'answer';
answer.append(root);
document.body.append(answer);
```

To run the JavaScript in a web browser a HTML page is needed.

> To get the `createModule` function we will import the `newtonraphsonwasm.js` with a script tag.

Reviewer: ```suggestion
To be able to use the `createModule` function, we will import the `newtonraphsonwasm.js` with a script tag.
```

```{.html file=src/js/example.html}
<!doctype html>
<!-- this HTML page is stored as src/js/example.html -->
<html>
  <script type="text/javascript" src="newtonraphsonwasm.js"></script>
  <script>
    <<wasmpromise>>
  </script>
</html>
```

The web browser can only load the `newtonraphsonwasm.js` file when hosted by a web server.

> Python ships with a built-in web server, we will use it to host the all files of the repository.

Reviewer: ```suggestion
Python ships with a built-in web server, we will use it to host the all files of the repository. The command below will start a web server and serve the page on port 8000.
```

Author: Solved slightly shorter

```{.awk #host-files}
python3 -m http.server 8000
```

> Visit [http://localhost:8000/src/js/example.html](http://localhost:8000/src/js/example.html) to see the root answer.

Reviewer: ```suggestion
Visit [http://localhost:8000/src/js/example.html](http://localhost:8000/src/js/example.html) to see the result of the calculation.
```

> The root finding answer was calculated using the C++ algorithm compiled to a WebAssembly module, executed by some JavaScript and rendered on a HTML page.

Reviewer: ```suggestion
The result of root finding was calculated using the C++ algorithm compiled to a WebAssembly module, executed by some JavaScript and rendered on a HTML page.
```

### Executing long running methods in JavaScript

> Executing a long running C++ method will block the browser from running any other code like updating the user interface. This is not very nice for the user. To run the method in the background, [web workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers) have been defined. A web worker runs in its own thread and can be interacted with from JavaScript using messages.

Reviewer: ```suggestion
Executing a long running C++ method will block the browser from running any other code like updating the user interface. In order to avoid this, the method can be run in the background using [web workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers). A web worker runs in its own thread and can be interacted with from JavaScript using messages.
```

### Single page application

> In the [Web application](#web_application) chapter a whole new page was rendered by the server even for a small change. With the advent of more powerful JavaScript engines in browsers and JavaScript methods to fetch JSON documents from a web service, it is possible to render the page with JavaScript and fetch a small change from the web service and re-render a small part of the page with JavaScript. The application running in the browser is called a [single page application](https://en.wikipedia.org/wiki/Single-page_application) or SPA.

Reviewer: ```suggestion
In the [Web application](#web_application) chapter, a whole new page was rendered by the server even for a small change. With the advent of more powerful JavaScript engines in browsers and JavaScript methods to fetch JSON documents from a web service, it is possible to prevent that. [Single Page Applications](https://en.wikipedia.org/wiki/Single-page_application)(SPA) can render the page and fetch a small change from the web service and re-render a small part of the page with JavaScript.
```
